ffi/tests/firewood/merkle_compatibility_test.go

```diff
@@ -0,0 +1,442 @@
+
+		remainingChildren = append(remainingChildren, dropChild.children...)
+	}
+
+	tr.children = commitProposal.children
+}
+
+func (tr *tree) checkDBHash() {
+	// Get the root hash from both databases.
+	merkleRoot, err := tr.merkleDB.GetMerkleRoot(context.Background())
+	tr.require.NoError(err)
+
+	fwdRoot, err := tr.fwdDB.Root()
+	tr.require.NoError(err)
+
+	// Compare the root hashes.
+	tr.require.Equal(merkleRoot[:], fwdRoot)
+}
+
+func (tr *tree) createProposalOnProposal() {
+	pr := tr.selectRandomProposal()
+
+	if pr == nil {
+		return
+	}
+
+	batchSize := tr.rand.Intn(maxBatchSize) + 1 // ensure at least one key-value pair
+	keys, vals := tr.createRandomBatch(batchSize)
+
+	fwdPr := pr.fwdView
+	merkleView := pr.merkleView
+
+	fwdChildPr, err := fwdPr.Propose(keys, vals)
+	tr.require.NoError(err)
+
+	merkleViewChange := merkledb.ViewChanges{}
+	for i := range keys {
+		merkleViewChange.BatchOps = append(merkleViewChange.BatchOps, database.BatchOp{
+			Key:   keys[i],
+			Value: vals[i],
+		})
+	}
+	merkleChildView, err := merkleView.NewView(context.Background(), merkleViewChange)
+	tr.require.NoError(err)
+
+	fwdRoot, err := fwdChildPr.Root()
+	tr.require.NoError(err)
+	merkleRoot, err := merkleChildView.GetMerkleRoot(context.Background())
+	tr.require.NoError(err)
+	tr.require.Equal(fwdRoot, merkleRoot[:])
+
+	tr.nextID++
+	newProposal := &proposal{
+		parentID:   pr.id,
+		id:         tr.nextID,
+		merkleView: merkleChildView,
+		fwdView:    fwdChildPr,
+	}
+	pr.children = append(pr.children, newProposal)
+	tr.proposals[newProposal.id] = newProposal
+}
+
+func (tr *tree) createProposalOnDB() {
+	batchSize := tr.rand.Intn(maxBatchSize) + 1 // ensure at least one key-value pair
+	keys, vals := tr.createRandomBatch(batchSize)
+	fwdPr, err := tr.fwdDB.Propose(keys, vals)
+	tr.require.NoError(err)
+
+	merkleViewChange := merkledb.ViewChanges{}
+	for i := range keys {
+		merkleViewChange.BatchOps = append(merkleViewChange.BatchOps, database.BatchOp{
+			Key:   keys[i],
+			Value: vals[i],
+		})
+	}
+	merkleChildView, err := tr.merkleDB.NewView(context.Background(), merkleViewChange)
+	tr.require.NoError(err)
+
+	fwdRoot, err := fwdPr.Root()
+	tr.require.NoError(err)
+	merkleRoot, err := merkleChildView.GetMerkleRoot(context.Background())
+	tr.require.NoError(err)
+	tr.require.Equal(fwdRoot, merkleRoot[:])
+
+	tr.nextID++
+	newProposal := &proposal{
+		parentID:   tr.id,
+		id:         tr.nextID,
+		merkleView: merkleChildView,
+		fwdView:    fwdPr,
+	}
+	tr.children = append(tr.children, newProposal)
+	tr.proposals[newProposal.id] = newProposal
+}
+
+func (tr *tree) proposalGet() {
+	pr := tr.selectRandomProposal()
+	if pr == nil {
+		return
+	}
+
+	keys := tr.selectRandomKeys(maxBatchSize)
+	for _, key := range keys {
+		fwdVal, fwdErr := pr.fwdView.Get(key)
+		merkleVal, merkleErr := pr.merkleView.GetValue(context.Background(), key)
+
+		// XXX: handling of non-existent keys diverges between merkleDB and firewood
```

Comment:
same comment and code earlier; can this be refactored into a helper?

Reply:
We could factor even more out more easily if we adopted a View interface, so that we could pass in different view types and execute the same test logic, but for now I will add a helper function to perform the assertion on results from firewood/merkle